File: phases/16-software-engineering-and-architecture/04-gof-patterns-that-still-matter/code/main.ts
/**
 * GoF Patterns That Still Matter — TypeScript Implementations
 * Phase 16 — Software Engineering & Architecture
 *
 * Seven patterns, each self-contained and runnable.
 * Run: npx tsx main.ts   (or: ts-node main.ts)
 */

// ── 1. Observer ──────────────────────────────────────────────────────────────

type Listener<T> = (data: T) => void;
type Unsubscribe = () => void;

class EventEmitter {
  private listeners = new Map<string, Set<Listener<unknown>>>();

  on<T>(event: string, fn: Listener<T>): Unsubscribe {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, new Set());
    }
    const set = this.listeners.get(event)!;
    set.add(fn as Listener<unknown>);
    return () => {
      set.delete(fn as Listener<unknown>);
      if (set.size === 0) this.listeners.delete(event);
    };
  }

  once<T>(event: string, fn: Listener<T>): Unsubscribe {
    const wrapper: Listener<T> = (data) => {
      unsub();
      fn(data);
    };
    const unsub = this.on(event, wrapper);
    return unsub;
  }

  emit<T>(event: string, data: T): void {
    this.listeners.get(event)?.forEach((fn) => fn(data));
  }
}

function demoObserver(): void {
  console.log("=== Observer ===");
  const bus = new EventEmitter();
  const unsubA = bus.on<number>("price", (p) =>
    console.log(`  Dashboard A: ${p}`)
  );
  bus.on<number>("price", (p) =>
    console.log(`  Dashboard B: ${p}`)
  );
  console.log("  First emit:");
  bus.emit("price", 142.5);
  unsubA();
  console.log("  After unsubscribing A:");
  bus.emit("price", 143.0);
}

// ── 2. Strategy ──────────────────────────────────────────────────────────────

interface PaymentStrategy {
  pay(amount: number): string;
}

class CreditCardPayment implements PaymentStrategy {
  constructor(private cardNumber: string) {}
  pay(amount: number): string {
    return `  Charged $${amount.toFixed(2)} to card ending ${this.cardNumber.slice(-4)}`;
  }
}

class CryptoPayment implements PaymentStrategy {
  constructor(private wallet: string) {}
  pay(amount: number): string {
    return `  Transferred $${amount.toFixed(2)} from wallet ${this.wallet.slice(0, 8)}`;
  }
}

class PayPalPayment implements PaymentStrategy {
  constructor(private email: string) {}
  pay(amount: number): string {
    return `  Sent $${amount.toFixed(2)} via PayPal to ${this.email}`;
  }
}

class PaymentProcessor {
  private strategy: PaymentStrategy;

  constructor(strategy: PaymentStrategy) {
    this.strategy = strategy;
  }

  setStrategy(strategy: PaymentStrategy): void {
    this.strategy = strategy;
  }

  process(amount: number): string {
    return this.strategy.pay(amount);
  }
}

function demoStrategy(): void {
  console.log("\n=== Strategy ===");
  const processor = new PaymentProcessor(
    new CreditCardPayment("4242424242424242")
  );
  console.log(processor.process(99.99));
  processor.setStrategy(new CryptoPayment("0xABCDEF1234567890"));
  console.log(processor.process(49.99));
  processor.setStrategy(new PayPalPayment("user@example.com"));
  console.log(processor.process(24.99));
}

// ── 3. Command (with Undo) ──────────────────────────────────────────────────

interface Command {
  execute(): void;
  undo(): void;
}

class AddTextCommand implements Command {
  constructor(
    private editor: TextEditor,
    private text: string,
    private position: number
  ) {}

  execute(): void {
    const content = this.editor.content;
    this.editor.content =
      content.slice(0, this.position) + this.text + content.slice(this.position);
  }

  undo(): void {
    const content = this.editor.content;
    const end = this.position + this.text.length;
    this.editor.content = content.slice(0, this.position) + content.slice(end);
  }
}

class DeleteTextCommand implements Command {
  private deleted = "";

  constructor(
    private editor: TextEditor,
    private length: number,
    private position: number
  ) {}

  execute(): void {
    const content = this.editor.content;
    this.deleted = content.slice(this.position, this.position + this.length);
    this.editor.content =
      content.slice(0, this.position) +
      content.slice(this.position + this.length);
  }

  undo(): void {
    const content = this.editor.content;
    this.editor.content =
      content.slice(0, this.position) +
      this.deleted +
      content.slice(this.position);
  }
}

class TextEditor {
  content = "";
  private history: Command[] = [];
  private redoStack: Command[] = [];

  execute(command: Command): void {
    command.execute();
    this.history.push(command);
    this.redoStack = [];
  }

  undo(): boolean {
    const cmd = this.history.pop();
    if (!cmd) return false;
    cmd.undo();
    this.redoStack.push(cmd);
    return true;
  }

  redo(): boolean {
    const cmd = this.redoStack.pop();
    if (!cmd) return false;
    cmd.execute();
    this.history.push(cmd);
    return true;
  }
}

function demoCommand(): void {
  console.log("\n=== Command ===");
  const editor = new TextEditor();
  editor.execute(new AddTextCommand(editor, "Hello", 0));
  console.log(`  After 'Hello': '${editor.content}'`);
  editor.execute(new AddTextCommand(editor, " World", 5));
  console.log(`  After ' World': '${editor.content}'`);
  editor.undo();
  console.log(`  After undo: '${editor.content}'`);
  editor.redo();
  console.log(`  After redo: '${editor.content}'`);
  editor.execute(new DeleteTextCommand(editor, 5, 5));
  console.log(`  After delete: '${editor.content}'`);
  editor.undo();
  console.log(`  After undo delete: '${editor.content}'`);
}

// ── 4. Iterator ──────────────────────────────────────────────────────────────

class TreeNode<T> {
  constructor(
    public value: T,
    public left: TreeNode<T> | null = null,
    public right: TreeNode<T> | null = null
  ) {}
}

function* inOrder<T>(node: TreeNode<T> | null): Generator<T> {
  if (node === null) return;
  yield* inOrder(node.left);
  yield node.value;
  yield* inOrder(node.right);
}

function* preOrder<T>(node: TreeNode<T> | null): Generator<T> {
  if (node === null) return;
  yield node.value;
  yield* preOrder(node.left);
  yield* preOrder(node.right);
}

function demoIterator(): void {
  console.log("\n=== Iterator ===");
  const root = new TreeNode(
    4,
    new TreeNode(2, new TreeNode(1), new TreeNode(3)),
    new TreeNode(6, new TreeNode(5), new TreeNode(7))
  );
  console.log(`  In-order:  [${[...inOrder(root)].join(", ")}]`);
  console.log(`  Pre-order: [${[...preOrder(root)].join(", ")}]`);

  // Using the iterator manually
  const it = inOrder(root);
  let result = it.next();
  const firstThree: number[] = [];
  while (!result.done && firstThree.length < 3) {
    firstThree.push(result.value);
    result = it.next();
  }
  console.log(`  First 3 in-order: [${firstThree.join(", ")}]`);
}

// ── 5. Factory Method ────────────────────────────────────────────────────────

interface Notification {
  send(message: string): string;
}

interface NotificationFactory {
  create(): Notification;
}

class EmailNotification implements Notification {
  send(message: string): string {
    return `  Email sent: ${message}`;
  }
}

class SMSNotification implements Notification {
  send(message: string): string {
    return `  SMS sent: ${message}`;
  }
}

class PushNotification implements Notification {
  send(message: string): string {
    return `  Push notification: ${message}`;
  }
}

class EmailFactory implements NotificationFactory {
  create(): Notification {
    return new EmailNotification();
  }
}

class SMSFactory implements NotificationFactory {
  create(): Notification {
    return new SMSNotification();
  }
}

class PushFactory implements NotificationFactory {
  create(): Notification {
    return new PushNotification();
  }
}

class NotificationService {
  constructor(private factory: NotificationFactory) {}

  notify(message: string): string {
    const notification = this.factory.create();
    return notification.send(message);
  }
}

function demoFactory(): void {
  console.log("\n=== Factory Method ===");
  const factories: NotificationFactory[] = [
    new EmailFactory(),
    new SMSFactory(),
    new PushFactory(),
  ];
  for (const factory of factories) {
    const service = new NotificationService(factory);
    console.log(service.notify("Build succeeded!"));
  }
}

// ── 6. Decorator ─────────────────────────────────────────────────────────────

type AsyncFunction = (...args: unknown[]) => Promise<unknown>;

function logCalls<T extends AsyncFunction>(fn: T): T {
  const wrapper = async (...args: unknown[]) => {
    console.log(`  [LOG] Calling ${fn.name}`);
    const result = await fn(...args);
    console.log(`  [LOG] ${fn.name} returned: ${result}`);
    return result;
  };
  return wrapper as T;
}

function timeCalls<T extends AsyncFunction>(fn: T): T {
  const wrapper = async (...args: unknown[]) => {
    const start = performance.now();
    const result = await fn(...args);
    const elapsed = performance.now() - start;
    console.log(`  [TIME] ${fn.name} took ${elapsed.toFixed(2)}ms`);
    return result;
  };
  return wrapper as T;
}

function cacheCalls<T extends AsyncFunction>(fn: T): T {
  const cache = new Map<string, unknown>();
  const wrapper = async (...args: unknown[]) => {
    const key = JSON.stringify(args);
    if (cache.has(key)) {
      console.log(`  [CACHE] Hit for ${fn.name}(${args.join(", ")})`);
      return cache.get(key);
    }
    console.log(`  [CACHE] Miss for ${fn.name}(${args.join(", ")})`);
    const result = await fn(...args);
    cache.set(key, result);
    return result;
  };
  return wrapper as T;
}

// Build the decorator chain
const expensiveCompute = cacheCalls(
  timeCalls(
    logCalls(async function expensiveCompute(n: number): Promise<number> {
      await new Promise((r) => setTimeout(r, 10));
      return n * n;
    })
  )
);

async function demoDecorator(): Promise<void> {
  console.log("\n=== Decorator ===");
  console.log(`  Result: ${await expensiveCompute(5)}`);
  console.log(`  Result (cached): ${await expensiveCompute(5)}`);
  console.log(`  Result (new arg): ${await expensiveCompute(7)}`);
}

// ── 7. Adapter ───────────────────────────────────────────────────────────────

interface UserRepository {
  getUser(id: string): { id: string; name: string; email: string };
}

interface LegacyUser {
  user_id: string;
  first_name: string;
  last_name: string;
  email_address: string;
}

class LegacyAuthService {
  private db: Record<string, LegacyUser> = {
    "1": { user_id: "1", first_name: "Jane", last_name: "Doe", email_address: "jane@legacy.com" },
    "2": { user_id: "2", first_name: "John", last_name: "Smith", email_address: "john@legacy.com" },
  };

  fetchUser(uid: string): LegacyUser {
    return this.db[uid];
  }
}

class LegacyUserAdapter implements UserRepository {
  constructor(private legacyService: LegacyAuthService) {}

  getUser(id: string): { id: string; name: string; email: string } {
    const legacy = this.legacyService.fetchUser(id);
    return {
      id: legacy.user_id,
      name: `${legacy.first_name} ${legacy.last_name}`,
      email: legacy.email_address,
    };
  }
}

class ModernUserService {
  constructor(private repo: UserRepository) {}

  greet(userId: string): string {
    const user = this.repo.getUser(userId);
    return `  Hello, ${user.name} (${user.email})`;
  }
}

function demoAdapter(): void {
  console.log("\n=== Adapter ===");
  const legacy = new LegacyAuthService();
  const adapter = new LegacyUserAdapter(legacy);
  const service = new ModernUserService(adapter);
  console.log(service.greet("1"));
  console.log(service.greet("2"));
}

// ── Main ─────────────────────────────────────────────────────────────────────

async function main(): Promise<void> {
  demoObserver();
  demoStrategy();
  demoCommand();
  demoIterator();
  demoFactory();
  await demoDecorator();
  demoAdapter();
}

main();

File: phases/16-software-engineering-and-architecture/04-gof-patterns-that-still-matter/code/main.py
"""
GoF Patterns That Still Matter — Python Implementations
Phase 16 — Software Engineering & Architecture

Seven patterns, each self-contained and runnable.
Run: python main.py
"""

from __future__ import annotations

import functools
import time
from dataclasses import dataclass, field
from typing import Protocol, Generator


# ── 1. Observer ──────────────────────────────────────────────────────────────

class EventEmitter:
    """Observer pattern with unsubscribe support."""

    def __init__(self) -> None:
        self._listeners: dict[str, list[callable]] = {}

    def on(self, event: str, callback: callable) -> callable:
        """Subscribe and return an unsubscribe function."""
        self._listeners.setdefault(event, []).append(callback)

        def unsubscribe() -> None:
            listeners = self._listeners.get(event, [])
            if callback in listeners:
                listeners.remove(callback)

        return unsubscribe

    def emit(self, event: str, *args, **kwargs) -> None:
        """Notify all listeners of an event."""
        for callback in list(self._listeners.get(event, [])):
            callback(*args, **kwargs)


def demo_observer() -> None:
    print("=== Observer ===")
    bus = EventEmitter()
    unsub_a = bus.on("price", lambda p: print(f"  Dashboard A: {p}"))
    bus.on("price", lambda p: print(f"  Dashboard B: {p}"))
    print("  First emit:")
    bus.emit("price", 142.50)
    unsub_a()
    print("  After unsubscribing A:")
    bus.emit("price", 143.00)


# ── 2. Strategy ──────────────────────────────────────────────────────────────

class PaymentStrategy(Protocol):
    def pay(self, amount: float) -> str: ...


@dataclass
class CreditCardPayment:
    card_number: str

    def pay(self, amount: float) -> str:
        return f"  Charged ${amount:.2f} to card ending {self.card_number[-4:]}"


@dataclass
class CryptoPayment:
    wallet: str

    def pay(self, amount: float) -> str:
        return f"  Transferred ${amount:.2f} from wallet {self.wallet[:8]}"


@dataclass
class PayPalPayment:
    email: str

    def pay(self, amount: float) -> str:
        return f"  Sent ${amount:.2f} via PayPal to {self.email}"


class PaymentProcessor:
    """Context that uses a strategy swappable at runtime."""

    def __init__(self, strategy: PaymentStrategy) -> None:
        self._strategy = strategy

    def set_strategy(self, strategy: PaymentStrategy) -> None:
        self._strategy = strategy

    def process(self, amount: float) -> str:
        return self._strategy.pay(amount)


def demo_strategy() -> None:
    print("\n=== Strategy ===")
    processor = PaymentProcessor(CreditCardPayment("4242424242424242"))
    print(processor.process(99.99))
    processor.set_strategy(CryptoPayment("0xABCDEF1234567890"))
    print(processor.process(49.99))
    processor.set_strategy(PayPalPayment("user@example.com"))
    print(processor.process(24.99))


# ── 3. Command (with Undo) ──────────────────────────────────────────────────

class Command:
    """Base command with execute and undo."""

    def execute(self) -> None:
        raise UnsupportedOperation

    def undo(self) -> None:
        raise UnsupportedOperation


class AddTextCommand(Command):
    def __init__(self, editor: "TextEditor", text: str, position: int) -> None:
        self._editor = editor
        self._text = text
        self._position = position

    def execute(self) -> None:
        content = self._editor.content
        self._editor.content = content[: self._position] + self._text + content[self._position :]

    def undo(self) -> None:
        content = self._editor.content
        end = self._position + len(self._text)
        self._editor.content = content[: self._position] + content[end:]


class DeleteTextCommand(Command):
    def __init__(self, editor: "TextEditor", length: int, position: int) -> None:
        self._editor = editor
        self._length = length
        self._position = position
        self._deleted = ""

    def execute(self) -> None:
        content = self._editor.content
        self._deleted = content[self._position : self._position + self._length]
        self._editor.content = content[: self._position] + content[self._position + self._length :]

    def undo(self) -> None:
        content = self._editor.content
        self._editor.content = content[: self._position] + self._deleted + content[self._position :]


class TextEditor:
    """Editor with undo/redo stack."""

    def __init__(self) -> None:
        self.content = ""
        self._history: list[Command] = []
        self._redo_stack: list[Command] = []

    def execute(self, command: Command) -> None:
        command.execute()
        self._history.append(command)
        self._redo_stack.clear()

    def undo(self) -> bool:
        if not self._history:
            return False
        cmd = self._history.pop()
        cmd.undo()
        self._redo_stack.append(cmd)
        return True

    def redo(self) -> bool:
        if not self._redo_stack:
            return False
        cmd = self._redo_stack.pop()
        cmd.execute()
        self._history.append(cmd)
        return True


def demo_command() -> None:
    print("\n=== Command ===")
    editor = TextEditor()
    editor.execute(AddTextCommand(editor, "Hello", 0))
    print(f"  After 'Hello': '{editor.content}'")
    editor.execute(AddTextCommand(editor, " World", 5))
    print(f"  After ' World': '{editor.content}'")
    editor.undo()
    print(f"  After undo: '{editor.content}'")
    editor.redo()
    print(f"  After redo: '{editor.content}'")
    editor.execute(DeleteTextCommand(editor, 5, 5))
    print(f"  After delete: '{editor.content}'")
    editor.undo()
    print(f"  After undo delete: '{editor.content}'")


# ── 4. Iterator ──────────────────────────────────────────────────────────────

class TreeNode:
    def __init__(self, value: int, left: "TreeNode | None" = None, right: "TreeNode | None" = None):
        self.value = value
        self.left = left
        self.right = right


def in_order(node: TreeNode | None) -> Generator[int, None, None]:
    """Iterator pattern via Python generator — in-order traversal."""
    if node is None:
        return
    yield from in_order(node.left)
    yield node.value
    yield from in_order(node.right)


def pre_order(node: TreeNode | None) -> Generator[int, None, None]:
    if node is None:
        return
    yield node.value
    yield from pre_order(node.left)
    yield from pre_order(node.right)


def demo_iterator() -> None:
    print("\n=== Iterator ===")
    root = TreeNode(4, TreeNode(2, TreeNode(1), TreeNode(3)), TreeNode(6, TreeNode(5), TreeNode(7)))
    print(f"  In-order:  {list(in_order(root))}")
    print(f"  Pre-order: {list(pre_order(root))}")


# ── 5. Factory Method ────────────────────────────────────────────────────────

class Notification(Protocol):
    def send(self, message: str) -> str: ...


class EmailNotification:
    def send(self, message: str) -> str:
        return f"  Email sent: {message}"


class SMSNotification:
    def send(self, message: str) -> str:
        return f"  SMS sent: {message}"


class PushNotification:
    def send(self, message: str) -> str:
        return f"  Push notification: {message}"


class NotificationFactory(Protocol):
    def create(self) -> Notification: ...


class EmailFactory:
    def create(self) -> Notification:
        return EmailNotification()


class SMSFactory:
    def create(self) -> Notification:
        return SMSNotification()


class PushFactory:
    def create(self) -> Notification:
        return PushNotification()


class NotificationService:
    """Uses a factory to create notifications — doesn't know the concrete type."""

    def __init__(self, factory: NotificationFactory) -> None:
        self._factory = factory

    def notify(self, message: str) -> str:
        notification = self._factory.create()
        return notification.send(message)


def demo_factory() -> None:
    print("\n=== Factory Method ===")
    for factory_cls in [EmailFactory, SMSFactory, PushFactory]:
        service = NotificationService(factory_cls())
        print(service.notify("Build succeeded!"))


# ── 6. Decorator ─────────────────────────────────────────────────────────────

def log_calls(fn: callable) -> callable:
    @functools.wraps(fn)
    def wrapper(*args, **kwargs):
        print(f"  [LOG] Calling {fn.__name__}")
        result = fn(*args, **kwargs)
        print(f"  [LOG] {fn.__name__} returned: {result}")
        return result

    return wrapper


def time_calls(fn: callable) -> callable:
    @functools.wraps(fn)
    def wrapper(*args, **kwargs):
        start = time.monotonic()
        result = fn(*args, **kwargs)
        elapsed = time.monotonic() - start
        print(f"  [TIME] {fn.__name__} took {elapsed:.4f}s")
        return result

    return wrapper


def cache_calls(fn: callable) -> callable:
    _cache: dict[tuple, object] = {}

    @functools.wraps(fn)
    def wrapper(*args):
        if args not in _cache:
            _cache[args] = fn(*args)
            print(f"  [CACHE] Miss for {fn.__name__}{args}")
        else:
            print(f"  [CACHE] Hit for {fn.__name__}{args}")
        return _cache[args]

    return wrapper


@log_calls
@time_calls
@cache_calls
def expensive_compute(n: int) -> int:
    """Simulate an expensive computation."""
    time.sleep(0.01)
    return n * n


def demo_decorator() -> None:
    print("\n=== Decorator ===")
    print(f"  Result: {expensive_compute(5)}")
    print(f"  Result (cached): {expensive_compute(5)}")
    print(f"  Result (new arg): {expensive_compute(7)}")


# ── 7. Adapter ───────────────────────────────────────────────────────────────

class UserRepository(Protocol):
    def get_user(self, user_id: str) -> dict: ...


@dataclass
class LegacyUser:
    user_id: str
    first_name: str
    last_name: str
    email_address: str


class LegacyAuthService:
    """A legacy system we can't modify."""

    def fetch_user(self, uid: str) -> LegacyUser:
        _db = {
            "1": LegacyUser("1", "Jane", "Doe", "jane@legacy.com"),
            "2": LegacyUser("2", "John", "Smith", "john@legacy.com"),
        }
        return _db[uid]


class LegacyUserAdapter:
    """Adapts LegacyAuthService to UserRepository interface."""

    def __init__(self, legacy_service: LegacyAuthService) -> None:
        self._service = legacy_service

    def get_user(self, user_id: str) -> dict:
        legacy = self._service.fetch_user(user_id)
        return {
            "id": legacy.user_id,
            "name": f"{legacy.first_name} {legacy.last_name}",
            "email": legacy.email_address,
        }


class ModernUserService:
    """The new system our code expects to work with."""

    def __init__(self, repo: UserRepository) -> None:
        self._repo = repo

    def greet(self, user_id: str) -> str:
        user = self._repo.get_user(user_id)
        return f"  Hello, {user['name']} ({user['email']})"


def demo_adapter() -> None:
    print("\n=== Adapter ===")
    legacy = LegacyAuthService()
    adapter = LegacyUserAdapter(legacy)
    service = ModernUserService(adapter)
    print(service.greet("1"))
    print(service.greet("2"))


# ── Main ─────────────────────────────────────────────────────────────────────

def main() -> None:
    demo_observer()
    demo_strategy()
    demo_command()
    demo_iterator()
    demo_factory()
    demo_decorator()
    demo_adapter()


if __name__ == "__main__":
    main()